File: cpp/ctci/is_unique.cc
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

// Version with no bit vector
bool is_unique(string input) {
  if (input.size() > 128) return false;

  unordered_map<int, bool> ht = {};
  for (int i = 0; i < input.size(); i++) {
    char cur_char = int(input[i]);
    if (ht.find(cur_char) != ht.end())
      return false;
    ht[cur_char] = true;
  }
  return true;
}

// Version with bit vector
bool is_unique_2(string input) {
  // Try using a bit vector to keep track
  int tester = 0;
  for (int i = 0; i < input.size(); i++) {
    int cur_char =  1 << int(input[i] - 'a');
    if (tester & cur_char) return false;
    tester |= cur_char;
  }
  return true;
}

int main () {
    string test = "abcde";
    cout << is_unique(test) << endl;
    test = "abcdea";
    cout << is_unique(test) << endl;

    test = "abcde";
    cout << is_unique_2(test) << endl;
    test = "abcdea";
    cout << is_unique_2(test) << endl;
    return 0;
}

File: python/ctci/arrays_and_strings/is_unique.py
def is_unique(in_str):
    ht = dict()
    for char in in_str:
        # Check if char is in the ht
        if not ord(char) in ht:
            ht[ord(char)] = 1
        else:
            return False
    return True

def is_unique_bit(in_str):
    bv = int(0)
    for char in in_str:
        char_bit = 1 << (ord(char) - ord('a'))
        if bv & char_bit:
            return False
        bv |= char_bit
    return True

if __name__ == '__main__':
    test = 'abcde'
    print(is_unique(test))
    test = 'abcdea'
    print(is_unique(test))

    test = 'abcde'
    print(is_unique_bit(test))
    test = 'abcdea'
    print(is_unique_bit(test))
